design/cdk-assets.md

```diff
@@ -0,0 +1,244 @@
+  readonly bucketName: string;
+  readonly objectKey: string;
+}
+
+interface ImageAssetSource {
+  readonly directory: string;                           // docker build context directory
+  readonly dockerBuildArgs?: { [arg: string]: string }; // optional args to "docker build"
+  readonly dockerBuildTarget?: string;                  // docker build --target to use
+  readonly dockerFile?: string;                         // custom name for Dockerfile
+}
+
+interface ImageAssetDestination {
+  readonly assumeRoleArn: string;                       // iam role to assume
+  readonly assumeRoleExternalId?: string;               // external id to pass to assume-role
+  readonly repositoryName: string;                      // ECR repository name
+  readonly imageName: string;                           // image tag to use
+}
+```
+
+Example of `assets.json` with two assets: a docker image and a file asset. Both assets are published to two destinations (S3/ECR).
+
+```json
+{
+  "version": "assets-1.0",
+  "images": {
+    "d31ca1aef8d1b68217852e7aea70b1e857d107b47637d5160f9f9a1b24882d2a": {
+      "source": {
+        "packaging": "docker",
+        "directory": "my-image",
+        "dockerFile": "CustomDockerFile",
+        "dockerBuildArgs": { "label": "prod" },
+        "dockerBuildTarget": "my-target"
+      },
+      "destinations": [
+        {
+          "repositoryName": "aws-cdk-images-2222222222US-us-east-1",
+          "imageName": "d31ca1aef8d1b68217852e7aea70b1e857d107b47637d5160f9f9a1b24882d2a",
+          "assumeRoleArn": "arn:aws:iam::2222222222US:role/aws-cdk-publish-2222222222US-us-east-1"
+        },
+        {
+          "repositoryName": "aws-cdk-images-3333333333EU-eu-west-2",
+          "imageName": "d31ca1aef8d1b68217852e7aea70b1e857d107b47637d5160f9f9a1b24882d2a",
+          "assumeRoleArn": "arn:aws:iam::3333333333EU:role/aws-cdk-publish-3333333333EU-eu-west-2"
+        }
+      ]
+    }
+  },
+  "files": {
+    "a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57": {
+      "source": {
+        "packaging": "zip",
+        "file": "myzipdirectory"
+      },
+      "destinations": [
+        {
+          "bucketName": "aws-cdk-files-2222222222US-us-east-1",
+          "objectKey": "a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57.zip"
+        },
+        {
+          "bucketName": "aws-cdk-files-3333333333EU-us-west-2",
+          "objectKey": "a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57.zip",
+          "assumeRoleArn": "arn:aws:iam::3333333333EU:role/aws-cdk-publish-3333333333EU-eu-west-2"
+        }
+      ]
+    }
+  }
+}
+```
+
+## API
+
+`cdk-assets` is designed as a stand-alone command line program and a library, so it can be integrated into other tools such as the CDK CLI or executed individually as a step in a CI/CD pipeline.
+
+### `publish`
+
+```shell
+cdk-assets publish DIR [ASSET-ID,ASSET-ID...]
```

Comment:
Nice to have. We will add additional features in the future as needed. I don't see a specific need right now. Do you?

Reply:
I have a feeling we'll need the `-f` option right out the door.

Users may want to force publish any assets they suspect might be altered out of band.

Reply:
That's fine. This spec tries to define the minimum required to support CI/CD. We can totally release this with additional capabilities.

Reply:
I was thinking of this a bit further. 

The expected state of the world for a user after running `cdk-assets publish` is that all of the assets are published with the expected contents. Correct me if I'm wrong here already?

If we use the existence of keys (for S3) and tags (for ECR) to skip a re-publish, it could mean that the artifacts could have been modified out-of-band, i.e., not the state of the world that the user expects.

Perhaps we should check beyond just keys and validate the content, such as using S3's ETag.

Reply:
There is an [implicit] assumption that users are not going to temper with deployment artifacts manually. It's true not only for assets but for any AWS resource maintained through code. Not even sure this requires stating but I will add a note.